### ISSUE TITLE
Time Zone offset required (for SQL queries)

### ISSUE THREAD

**Issue by doclucas** [thirtybees/thirtybees#581]
Currently inserting queries with time attributes such as `NOW()` inserts GMT time, regardless of the zone configuration settings in the BO.
A good example is the code of the cron manager:
`'updated_at' => ['type' => 'sql', 'value' => 'NOW()'],`
Which inserts the time from GMT, and then other functions in that file compare the time to the correct php time, which mean - some crons won't run, depends on what zone the store is set to.
https://forum.thirtybees.com/topic/1285/cron-jobs-last-execution/13

To make a long story short, there must be added configuration settings that properly set MySQL time to the time zone, which match the php time.
Some examples:
https://stackoverflow.com/questions/34428563/set-timezone-in-php-and-mysql

**Comment by doclucas** [comment]
It seems there is something incorrect in TB code.
@Traumflug 
Why does does this code updates NOW() as GMT time instead of the real MySQL time (even tho the BO time zone is set correctly, and is not GMT)?
```
                    Db::getInstance()->update(
                        \Cronjobs::TABLE,
                        [
                            'updated_at' => ['type' => 'sql', 'value' => 'NOW()'],
                            'active'     => ['type' => 'sql', 'value' => 'IF(`one_shot` = TRUE, FALSE, `active`)'],
                        ],
                        '`id_cronjob` = '.(int) $cron['id_cronjob']
                    );
```
`mysql -e "SELECT NOW();"` shows me the correct server time, not GMT time.
Where is it configured in TB application?

**Comment by Traumflug** [comment]
Command line, DB users and even DB connections can each have their own time zone:

https://dev.mysql.com/doc/refman/8.0/en/time-zone-support.html
https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_now

**Comment by doclucas** [comment]
Thanks, @Traumflug 
Unfortunately that doesn't answer my question.
I made 2 test simple database connections scripts (php) in the same TB folder.
One script is mysqli, the other is PDO. Both use TB username and password and connect to the same TB database.
In both cases I got the correct time returned for `now()`, ie the timezone is correct. When using TB I am getting GMT time for `now()`.
The question is - where in TB code is being set/reset?
Also - what connection type does TB use? PDO or MySQLi?

**Comment by doclucas** [comment]
Exactly.
One can workaround the bug by simply adding a hardcoded
`@date_default_timezone_set('MY-TIMEZONE');` to the top of config.inc.php, but that's obviously not a solution for the core.

I wonder if SQL timezone is being set correctly on the PS 1.6 version  that TB is based on, and if so, that means that TB changed the initialization sequence of some important code, which may have caused more issues we may or may not be aware of right now.

I wish I knew the ins and outs of TB code well enough to be confident to try to fix this myself, but I am still new around here.
@getdatakick do you think you can get this challenging issue resolved safely?
or @Traumflug ?

**Comment by Traumflug** [comment]
A quick grep brings up quite a number of timezone changes:

```none
install-dev/classes/controllerHttp.php:147:            @date_default_timezone_set($session->shopTimezone);
install-dev/classes/controllerConsole.php:131:        @date_default_timezone_set($datas->timezone);
install-dev/init.php:88:    @date_default_timezone_set('UTC');
config/config.inc.php:152:@date_default_timezone_set(Configuration::get('PS_TIMEZONE'));
controllers/admin/AdminLocalizationController.php:161:        if (function_exists('date_default_timezone_set')) {
```

`PS_TIMEZONE` is apparently stored in the database, table `configuration`:
```none
classes/Configuration.php:99:    const TIMEZONE = 'PS_TIMEZONE';
install-dev/models/install.php:637:        Configuration::updateGlobalValue('PS_TIMEZONE', $data['shopTimezone']);
install-dev/data/xml/configuration.xml:185:    <configuration id="PS_TIMEZONE" name="PS_TIMEZONE">
config/config.inc.php:152:@date_default_timezone_set(Configuration::get('PS_TIMEZONE'));
controllers/admin/AdminLocalizationController.php:162:            $this->fields_options['general']['fields']['PS_TIMEZONE'] = [
```
*AdminLocalizationController* is the back office page menu -> *Localization* -> *Localization* and can be changed there.

**Comment by getdatakick** [comment]
I can prepare fix for this issue

**Comment by getdatakick** [comment]
The fix is ready. @Traumflug could you please review commit d3e24a6ba4c0be3cc1526822d401b48d3acf847b

**Comment by Traumflug** [comment]
Looks good. Then added another commit to simplify the fallback in `Tools::getTimeZone()`.

https://github.com/thirtybees/thirtybees/commits/travis

**Comment by getdatakick** [comment]
Great. It seems like the fix breaks CI, though: https://travis-ci.org/thirtybees/thirtybees/jobs/421084246

It's not a problem with fix itself, but with testing/mock environments. The problem is with this file: [/tests/_support/override/classes/db/Db.php](https://github.com/thirtybees/thirtybees/blob/1.0.x/tests/_support/override/classes/db/Db.php), it needs to override the new abstract method ```setTimeZone()```

**Comment by Traumflug** [comment]
Thanks for the hint, I'll fix this.

**Comment by Dh42** [comment]
I think GMT is the preferred method for storing time, not using they system time. Using the system timezone will likely cause issues.

**Comment by doclucas** [comment]
Yes, it's not using the system timezone, it's using the timezone set in thirtybees BO.
php was correctly using the BO timezone, while SQL wasn't.
This beautiful patch has fixed it, Les.

**Comment by Dh42** [comment]
What I am saying is that shouldn't everything be stored with gmt time and then adjusted in script?

**Comment by doclucas** [comment]
If you mean database inserts and updates should be GMT and then fetched with timezone adjustments, then the answer is no, for many reasons, such as the entire logic of time-based modules has been written based on correct current time, which will require full rewrites of such modules etc. Also, I don't know any other ecommerce or other application that does it this way, they all do it like TB, but admittedly I don't know all of them.

**Comment by Dh42** [comment]
Doesn't this break when used with an array of slaves all around with world with their local times set?

**Comment by doclucas** [comment]
I am not sure I understand the question, but I can say for sure that Prestashop (also Magento and others) work this way as well and I don't see anything that broke there. afaik.

**Comment by Dh42** [comment]
My thought is what if your main write server is located in germany and the time is set to the local time. Then you have a slave located in brazil set to their local time. If you are not using gmt then you run into differences because of the time differences.

**Comment by techjp** [comment]
The best way to do it is as Lesley describes, everything stored as UTC and then conversions done for display. This way if a store changes its time zone or server time zone it will never end up with orders with future times or just times that aren’t correct for the new time zone.

However if everything for PS was designed around different time zones being used in the database then changing that will break every module that uses time. That would not be good...

**Comment by techjp** [comment]
Unfortunate that PS made such a bad decision about time and time zones. It’s very much industry standard to use UTC for storage but as you mention, changing it at this point would create a mess.

I guess the “solution” is for big sites to either do conversion themselves when they download data from their sites or to just run all their sites in UTC. Personally I’d opt for the former, given the choice.

**Comment by getdatakick** [comment]
> My thought is what if your main write server is located in germany and the time is set to the local time. Then you have a slave located in brazil set to their local time. If you are not using gmt then you run into differences because of the time differences.

No, because all db instances will be initialized to use the same timezone (the same as php server). So everything across the stack is synchronized, just not in UTC. 

Note that prestashop does not implement this synchronization, so they can actually be affected by this problem

**Comment by Traumflug** [comment]
> I think GMT is the preferred method for storing time, not using they system time. Using the system timezone will likely cause issues.
> [...]
> What I am saying is that shouldn't everything be stored with gmt time and then adjusted in script?

MySQL should take care of this. Having a connection set to a certain timezone and saving a time doesn't mean it gets stored internally with this timezone.

In other words: storing a time while being set to a European timezone and reading back this time gets the same time readout. No surprise on this. However, when switching to Brazil timezone or using a slave in the Brazil timezone, reading the very same stored time gives Brazil local time. Same stored data, different readout. Timezones aren't per database, but per database connection. MySQL converts as needed.

As @getdatakick pointed out, thirty bees sets all DB connections to the same timezone, no matter where the server is located or where the HTTP request comes from. That's the reason why one gets the same time readings all over the planet.

If the idea is to present local time readings to users, no problem from the database side. But a can of worms when trying to find out the user's preferred time and distinguishing user time readings and server time readings.
